### ISSUE TITLE
[BUG] Upgrade stucks in `Waiting for RKE2 to be upgraded`, fleet-agent was scaled down, but failed to scale up

### ISSUE THREAD

**Issue by w13915984028** [harvester/harvester#4119]
**Describe the bug**

![image](https://github.com/harvester/harvester/assets/31133476/79fff965-8b02-4cc6-88a5-1ae6183fc4cf)

`apply-manifests` finished , even with some error messages https://github.com/harvester/harvester/issues/4095#issuecomment-1599484467, but it did not stuck there 

```
logs of apply-manifests
....
addon.harvesterhci.io/vm-import-controller patched
addon.harvesterhci.io/pcidevices-controller patched
upgrade addon rancher_monitoring
upgrade addon rancher_logging

Tue Jun 20 20:50:22 UTC 2023
NAMESPACE     NAME                                          BUNDLEDEPLOYMENTS-READY   STATUS
fleet-local   fleet-agent-local                             1/1                       
fleet-local   local-managed-system-agent                    1/1                       
fleet-local   mcc-harvester                                 1/1                       
fleet-local   mcc-harvester-crd                             1/1                       
fleet-local   mcc-local-managed-system-upgrade-controller   1/1                       
fleet-local   mcc-rancher-logging-crd                       1/1                       
fleet-local   mcc-rancher-monitoring-crd                    1/1                       
all bundles are ready
Tue Jun 20 20:57:48 UTC 2023
```

```
harv31:/home/rancher # kk get upgrade.harvesterhci.io -A
NAMESPACE          NAME                 AGE
harvester-system   hvst-upgrade-scnqb   35m
harv31:/home/rancher # kk get upgrade.harvesterhci.io -A -oyaml
apiVersion: v1
items:
- apiVersion: harvesterhci.io/v1beta1
  kind: Upgrade
  metadata:
    creationTimestamp: "2023-06-20T20:29:00Z"
    finalizers:
    - wrangler.cattle.io/harvester-upgrade-controller
    generateName: hvst-upgrade-
    generation: 14
    labels:
      harvesterhci.io/latestUpgrade: "true"
      harvesterhci.io/upgradeState: UpgradingNodes
    name: hvst-upgrade-scnqb
    namespace: harvester-system
    resourceVersion: "56088"
    uid: a36e81b9-898f-4e5a-ade1-918994277f2a
  spec:
    image: ""
    logEnabled: true
    version: v1.2.0
  status:
    conditions:
    - status: Unknown
      type: Completed
    - lastUpdateTime: "2023-06-20T20:29:17Z"
      status: "True"
      type: LogReady
    - lastUpdateTime: "2023-06-20T20:29:50Z"
      status: "True"
      type: ImageReady
    - lastUpdateTime: "2023-06-20T20:30:43Z"
      status: "True"
      type: RepoReady
    - lastUpdateTime: "2023-06-20T20:37:17Z"
      status: "True"
      type: NodesPrepared
    - lastUpdateTime: "2023-06-20T20:50:29Z"
      status: "True"
      type: SystemServicesUpgraded
    - status: Unknown
      type: NodesUpgraded
    imageID: harvester-system/harvester-iso-lnrn4
    nodeStatuses:
      harv31:
        state: Images preloaded
    previousVersion: v1.1.2
    repoInfo: |
      release:
          harvester: ab3fdceb
          harvesterChart: 0.0.0-fix4071-2-ab3fdceb
          os: Harvester 9702369-dirty
          kubernetes: v1.25.9+rke2r1
          rancher: v2.7.5-rc5
          monitoringChart: 102.0.0+up40.1.2
    singleNode: harv31
    upgradeLog: hvst-upgrade-scnqb-upgradelog
kind: List
metadata:
  resourceVersion: ""
harv31:/home/rancher # 
```

log of single-node-upgrade
```
+ upgrade_rke2
++ mktemp -p /host/usr/local/upgrade_tmp
+ patch_file=/host/usr/local/upgrade_tmp/tmp.Yh3uoFcmaO
+ cat
+ kubectl patch clusters.provisioning.cattle.io local -n fleet-local --patch-file /host/usr/local/upgrade_tmp/tmp.Yh3uoFcmaO --type merge
cluster.provisioning.cattle.io/local patched
+ wait_rke2_upgrade
++ echo -n v1.25.9+rke2r1
++ sed 's/-rc[[:digit:]]*//g'
+ REPO_RKE2_VERSION_WITHOUT_RC=v1.25.9+rke2r1
++ get_node_rke2_version
++ kubectl get node harv31 -o yaml
++ yq -e e .status.nodeInfo.kubeletVersion -
Waiting for RKE2 to be upgraded...
+ '[' v1.24.11+rke2r1 = v1.25.9+rke2r1 ']'
+ echo 'Waiting for RKE2 to be upgraded...'
+ sleep 5

....

Waiting for RKE2 to be upgraded...
++ get_node_rke2_version
++ kubectl get node harv31 -o yaml
++ yq -e e .status.nodeInfo.kubeletVersion -
+ '[' v1.24.11+rke2r1 = v1.25.9+rke2r1 ']'
+ echo 'Waiting for RKE2 to be upgraded...'
+ sleep 5
Waiting for RKE2 to be upgraded...


harv31:~ # kk get node -A
NAME     STATUS   ROLES                              AGE   VERSION
harv31   Ready    control-plane,etcd,master,worker   32d   v1.24.11+rke2r1
harv31:~ # 


harv31:~ # kk get clusters.provisioning.cattle.io -A -oyaml
apiVersion: v1
items:
- apiVersion: provisioning.cattle.io/v1
  kind: Cluster
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"provisioning.cattle.io/v1","kind":"Cluster","metadata":{"annotations":{},"name":"local","namespace":"fleet-local"},"spec":{"kubernetesVersion":"v1.24.11+rke2r1","rkeConfig":{"controlPlaneConfig":null}}}
...
  spec:
    kubernetesVersion: v1.25.9+rke2r1

```

local-managed-system-upgrade-controller is waiting for upgrade
```
harv31:~ # kk get bundle -A
NAMESPACE     NAME                                          BUNDLEDEPLOYMENTS-READY   STATUS
fleet-local   fleet-agent-local                             1/1                       
fleet-local   local-managed-system-agent                    1/1                       
fleet-local   mcc-harvester                                 1/1                       
fleet-local   mcc-harvester-crd                             1/1                       
fleet-local   mcc-local-managed-system-upgrade-controller   0/1                       WaitApplied(1) [Cluster fleet-local/local]
fleet-local   mcc-rancher-logging-crd                       1/1                       
fleet-local   mcc-rancher-monitoring-crd                    1/1                       
harv31:~ # 
```

but `fleet-agent` pod is gone:
```
harv31:~ # kk get pods -A
NAMESPACE                  NAME                                                        READY   STATUS      RESTARTS      AGE
cattle-fleet-system        fleet-controller-56786984f4-lfhsf                           1/1     Running     0             50m
cattle-fleet-system        gitjob-845b9dcc47-kq967                                     1/1     Running     0             50m
```

`fleet-controller` has such error:

> time="2023-06-20T20:51:23Z" level=info msg="While calculating status.ResourceKey, error running helm template for bundle mcc-local-managed-system-upgrade-controller with target options from : chart requires kubeVersion: >= 1.23.0-0 which is incompatible with Kubernetes v1.20.0"

```
ime="2023-06-20T20:49:55Z" level=info msg="waiting for service account token key to be populated for secret fleet-local/import-token-local-cd85db21-109c-4b93-b7b8-a6b64d75f80c-token"
time="2023-06-20T20:49:57Z" level=info msg="Deleted old agent for cluster (fleet-local/local) in namespace cattle-fleet-local-system"
time="2023-06-20T20:49:58Z" level=info msg="Cluster import for 'fleet-local/local'. Deployed new agent"
time="2023-06-20T20:49:58Z" level=info msg="Deleted old agent for cluster (fleet-local/local) in namespace cattle-fleet-local-system"
time="2023-06-20T20:49:58Z" level=info msg="Cluster import for 'fleet-local/local'. Deployed new agent"
time="2023-06-20T20:50:00Z" level=info msg="Cluster registration request 'fleet-local/request-2drgz', cluster 'fleet-local/local' granted [false], creating cluster and request service account"
time="2023-06-20T20:50:00Z" level=info msg="Cluster registration request 'fleet-local/request-mdphm', cluster 'fleet-local/local' granted [false], creating cluster and request service account"
time="2023-06-20T20:50:00Z" level=info msg="waiting for service account token key to be populated for secret cluster-fleet-local-local-1a3d67d0a899/request-2drgz-c9636d69-579d-470f-879d-bced1069b906-token"
time="2023-06-20T20:50:00Z" level=info msg="waiting for service account token key to be populated for secret cluster-fleet-local-local-1a3d67d0a899/request-mdphm-7fa9c607-c7db-4ae0-9b0f-9aad3338d517-token"
time="2023-06-20T20:50:02Z" level=info msg="Cluster registration request 'fleet-local/request-2drgz', cluster 'fleet-local/local' granted [true], creating cluster and request service account"
time="2023-06-20T20:50:02Z" level=info msg="Cluster registration request 'fleet-local/request-2drgz', cluster 'fleet-local/local' granted [true], creating cluster and request service account"
time="2023-06-20T20:50:02Z" level=info msg="Cluster registration request 'fleet-local/request-mdphm', cluster 'fleet-local/local' granted [true], creating cluster and request service account"
time="2023-06-20T20:50:02Z" level=info msg="Cluster registration request 'fleet-local/request-mdphm', cluster 'fleet-local/local' granted [true], creating cluster and request service account"

time="2023-06-20T20:51:23Z" level=info msg="While calculating status.ResourceKey, error running helm template for bundle mcc-local-managed-system-upgrade-controller with target options from : chart requires kubeVersion: >= 1.23.0-0 which is incompatible with Kubernetes v1.20.0"

time="2023-06-20T20:54:09Z" level=info msg="Deleting orphaned content[s-882a70e702bd9ff5f694d7fb38f1c2781c3a77d01784b84dcc2f825c8201b]"
time="2023-06-20T20:54:09Z" level=info msg="Deleting orphaned content[s-2071566478ea38e5c6370fc4989578af40acc316792e36b067916258e03a8]"
time="2023-06-20T20:54:09Z" level=info msg="Deleting orphaned content[s-1c31172e241c5cdb02be20688ba3660435c0ce1a3b3cc56a63c56ac20d43c]"
time="2023-06-20T20:54:09Z" level=info msg="Deleting orphaned content[s-607b07bc72a1fc5fef261e81b989a5b5f0863e51730ccf22e9af6e5498ff7]"
time="2023-06-20T21:30:02Z" level=info msg="Deleting expired registration secret cattle-fleet-clusters-system/c-c8ee7f7a93be60d7cb174a44ed98b769791771be29cbd3b539806c9e7f3eb"
time="2023-06-20T21:30:02Z" level=info msg="Deleting expired registration secret cattle-fleet-clusters-system/c-108b8d9e610af40b419164367639af861293eb035d63723e978065ade031a"
```

**To Reproduce**
1. v1.1.2 -> v1.2.0 master-head  single-node cluster upgrade 

**Expected behavior**

upgrade is successful

**Support bundle**

**Environment**
 - Harvester ISO version: 
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630):

**Additional context**

Observed when testing https://github.com/harvester/harvester/pull/4117

Other related issues:
#4096 
#4095

**Comment by w13915984028** [comment]
This error from `fleet-controller` may be critical

> time="2023-06-20T20:51:23Z" level=info msg="While calculating status.ResourceKey, error running helm template for bundle mcc-local-managed-system-upgrade-controller with target options from : chart requires kubeVersion: >= 1.23.0-0 which is incompatible with Kubernetes v1.20.0"

`fleet-agent` POD is gone

```
harv31:~ # kk get pods -A | grep fleet
cattle-fleet-system        fleet-controller-56786984f4-lfhsf                           1/1     Running     0             68m
cattle-fleet-system        gitjob-845b9dcc47-kq967                                     1/1     Running     0             68m
harv31:~ # 
```

```
harv31:~ # kk get bundle -A
NAMESPACE     NAME                                          BUNDLEDEPLOYMENTS-READY   STATUS
fleet-local   fleet-agent-local                             1/1        
..           
fleet-local   mcc-local-managed-system-upgrade-controller   0/1                       WaitApplied(1) [Cluster fleet-local/local]
..             

harv31:~ # kk get bundledeployment -A
NAMESPACE                                NAME                                          DEPLOYED   MONITORED   STATUS
cluster-fleet-local-local-1a3d67d0a899   fleet-agent-local                             True       True        
..   
cluster-fleet-local-local-1a3d67d0a899   mcc-local-managed-system-upgrade-controller   True       True    
```

The `clusters.provisioning.cattle.io` is pending from `v1.24.11+rke2r1` -> `v1.25.9+rke2r1`

```
harv31:~ # kk get bundledeployment  -n cluster-fleet-local-local-1a3d67d0a899 fleet-agent-local -oyaml
apiVersion: fleet.cattle.io/v1alpha1
kind: BundleDeployment
metadata:
..
  name: fleet-agent-local
  namespace: cluster-fleet-local-local-1a3d67d0a899
  resourceVersion: "55726"
  uid: 85bd67dd-172f-459a-a85b-c66e389d1042
spec:
...
status:
  appliedDeploymentID: s-5a460ff0e462e12abaefdfc2f47fdedacf2330acb7a7d9d510b0dc5645f5c:32eb9f72e71f28d6a3ee815ddf77e461aa78ec0b6870afe4765c423d31bf6557
  conditions:
  - lastUpdateTime: "2023-05-19T09:13:41Z"
    status: "True"
    type: Installed
  - lastUpdateTime: "2023-06-20T20:50:06Z"
    status: "True"
    type: Deployed
  - lastUpdateTime: "2023-06-20T20:50:11Z"
    status: "True"
    type: Ready
  - lastUpdateTime: "2023-06-20T20:50:11Z"
    status: "True"
    type: Monitored
  display:
    deployed: "True"
    monitored: "True"
    state: Ready
  nonModified: true
  ready: true
  release: cattle-fleet-local-system/fleet-agent-local:12
  syncGeneration: 0
harv31:~ # 
```

**Comment by w13915984028** [comment]
**The old `fleet-agent` was scaled down to 0, and `fleet-controller` encounters error to deploy new `fleet-agent`.The upgrade hangs here.**

```
harv31:/home/rancher # kk describe deployment -n cattle-fleet-local-system fleet-agent

Replicas:               0 desired | 0 updated | 0 total | 0 available | 0 unavailable

Events:
  Type    Reason             Age   From                   Message
  ----    ------             ----  ----                   -------
  Normal  ScalingReplicaSet  60m   deployment-controller  Scaled down replica set fleet-agent-7b6499b8d6 to 0
```

```
time="2023-06-20T20:49:55Z" level=info msg="waiting for service account token key to be populated for secret fleet-local/import-token-local-cd85db21-109c-4b93-b7b8-a6b64d75f80c-token"
time="2023-06-20T20:49:57Z" level=info msg="Deleted old agent for cluster (fleet-local/local) in namespace cattle-fleet-local-system"
time="2023-06-20T20:49:58Z" level=info msg="Cluster import for 'fleet-local/local'. Deployed new agent"
time="2023-06-20T20:49:58Z" level=info msg="Deleted old agent for cluster (fleet-local/local) in namespace cattle-fleet-local-system"


time="2023-06-20T20:49:58Z" level=info msg="Cluster import for 'fleet-local/local'. Deployed new agent"
time="2023-06-20T20:50:00Z" level=info msg="Cluster registration request 'fleet-local/request-2drgz', cluster 'fleet-local/local' granted [false], creating cluster and request service account"
time="2023-06-20T20:50:00Z" level=info msg="Cluster registration request 'fleet-local/request-mdphm', cluster 'fleet-local/local' granted [false], creating cluster and request service account"
time="2023-06-20T20:50:00Z" level=info msg="waiting for service account token key to be populated for secret cluster-fleet-local-local-1a3d67d0a899/request-2drgz-c9636d69-579d-470f-879d-bced1069b906-token"
time="2023-06-20T20:50:00Z" level=info msg="waiting for service account token key to be populated for secret cluster-fleet-local-local-1a3d67d0a899/request-mdphm-7fa9c607-c7db-4ae0-9b0f-9aad3338d517-token"
time="2023-06-20T20:50:02Z" level=info msg="Cluster registration request 'fleet-local/request-2drgz', cluster 'fleet-local/local' granted [true], creating cluster and request service account"
time="2023-06-20T20:50:02Z" level=info msg="Cluster registration request 'fleet-local/request-2drgz', cluster 'fleet-local/local' granted [true], creating cluster and request service account"
time="2023-06-20T20:50:02Z" level=info msg="Cluster registration request 'fleet-local/request-mdphm', cluster 'fleet-local/local' granted [true], creating cluster and request service account"
time="2023-06-20T20:50:02Z" level=info msg="Cluster registration request 'fleet-local/request-mdphm', cluster 'fleet-local/local' granted [true], creating cluster and request service account"


time="2023-06-20T20:51:23Z" level=info msg="While calculating status.ResourceKey, error running helm template for bundle mcc-local-managed-system-upgrade-controller with target options from : chart requires kubeVersion: >= 1.23.0-0 which is incompatible with Kubernetes v1.20.0"
```

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.2.0-rc3`. Close this issue.

### Result
* The upgrade did not stuck in the `Waiting RKE2 to be upgraded` 

* We can pass the updating system stage and finish the `single` node upgrade to `v1.2.0-rc3`
  ![image](https://github.com/harvester/harvester/assets/29251855/47733a86-6ce0-45c3-9e50-d16a82f93d0e)

* And also pass the updating system stage and finish the `multi` node upgrade to `v1.2.0-rc3`
  ![image](https://github.com/harvester/harvester/assets/29251855/da35751b-6776-4100-9668-fef99b3b2467)

* Updating system jobs logs 
  ```
   wait for rollout of logging and audit                                                                                                                                                      │
  │ deployment "rancher-logging" successfully rolled out                                                                                                                                       │
  │ daemon set "rancher-logging-root-fluentbit" successfully rolled out                                                                                                                        │
  │ daemon set "rancher-logging-rke2-journald-aggregator" successfully rolled out                                                                                                              │
  │ daemon set "rancher-logging-kube-audit-fluentbit" successfully rolled out                                                                                                                  │
  │ success to wait rollout logging audit                                                                                                                                                      │
  │ there are 10 pods in cattle-logging-system after upgrade                                                                                                                                   │
  │ Logging Event Audit: finish converting                                                                                                                                                     │
  │ Applying extra manifests                                                                                                                                                                   │
  │ No extra manifests in /usr/local/share/extra_manifests/v1.1.2 to apply                                                                                                                     │
  │ addon.harvesterhci.io/vm-import-controller patched                                                                                                                                         │
  │ addon.harvesterhci.io/pcidevices-controller patched                                                                                                                                        │
  │ Error from server (NotFound): addons.harvesterhci.io "harvester-seeder" not found                                                                                                          │
  │ addon.harvesterhci.io/harvester-seeder created                                                                                                                                             │
  │ upgrade addon rancher_monitoring                                                                                                                                                           │
  │ upgrade addon rancher_logging                                                                                                                                                              │
  │                                                                                                                                                                                            │
  │ Tue Jul 11 14:49:34 UTC 2023                                                                                                                                                               │
  │ NAMESPACE     NAME                                          BUNDLEDEPLOYMENTS-READY   STATUS                                                                                               │
  │ fleet-local   fleet-agent-local                             1/1                                                                                                                            │
  │ fleet-local   local-managed-system-agent                    1/1                                                                                                                            │
  │ fleet-local   mcc-harvester                                 1/1                                                                                                                            │
  │ fleet-local   mcc-harvester-crd                             1/1                                                                                                                            │
  │ fleet-local   mcc-local-managed-system-upgrade-controller   1/1                                                                                                                            │
  │ fleet-local   mcc-rancher-logging-crd                       1/1                                                                                                                            │
  │ fleet-local   mcc-rancher-monitoring-crd                    1/1                                                                                                                            │
  │ all bundles are ready                                                                                                                                                                      │
  │ Stream closed EOF for harvester-system/hvst-upgrade-5vs5j-apply-manifests-kmc2g (apply) 
  ```

### Test Information
* Test Environment: single and four nodes harvester on provo bare machine
* Harvester version: v1.1.2 to v1.2.0-rc3

### Verify Steps
1. Prepare a single or four nodes v1.1.2 Harvester cluster
2. Create vlan network and create a virtual machine 
3. Upgrade to `v1.2.0-rc3`
4. Check the upgrade state
